server/src/lib.rs

```diff
@@ -186,25 +209,19 @@ impl Server {
 
     fn on_connect(
         &mut self,
-        conn: Result<quinn::NewConnection, quinn::ConnectionError>,
+        connection: quinn::Connection,
         mut send: mpsc::Sender<(ClientId, ClientEvent)>,
     ) {
-        let quinn::NewConnection {
-            connection,
-            uni_streams,
-            ..
-        } = match conn {
-            Ok(x) => x,
-            Err(e) => {
-                error!("incoming connection failed: {}", e);
-                return;
-            }
-        };
         let id = self.clients.insert(Client::new(connection.clone()));
         info!(id = ?id.0, address = %connection.remote_address(), "connection established");
         tokio::spawn(async move {
-            if let Err(e) = drive_recv(id, uni_streams, &mut send).await {
+            if let Err(e) = drive_recv(id, connection, &mut send).await {
+                // drive_recv returns an error when any connection-terminating issue occurs, so we
+                // send a Lost message to ensure the client is cleaned up. Note that this message may
+                // be redundant, as drive_recv also sends such messages.
```

Comment:
> drive_recv also sends such messages

Does it? Where?

Reply:
Whoops. It used to in an earlier WIP revision, but I switched to just closing the connection instead. I'll update this comment.